File: src/main/java/com/nibado/projects/advent/Day10.java
package com.nibado.projects.advent;

import static com.nibado.projects.advent.Util.printAnswer;
import static com.nibado.projects.advent.Util.readResourceAsString;

public class Day10 implements Runnable {

    @Override
    public void run() {
        String input = readResourceAsString("/day10.txt");

        printAnswer(apply(40, input));
        printAnswer(apply(50, input));
    }

    private static int apply(int times, String input) {
        for(int i = 0;i < times;i++) {
            input = next(input);
        }

        return input.length();
    }

    private static String next(String input) {
        StringBuilder builder = new StringBuilder();

        for(int i = 0, count = 0;i < input.length();) {
            while(i + count < input.length() && input.charAt(i) == input.charAt(i + count)) {
                count++;
            }
            builder.append(count).append(input.charAt(i));
            i += count;
            count = 0;
        }

        return builder.toString();
    }

    public static void main(String... argv) {
        new Day10().run();
    }
}

File: src/main/scala/com/nibado/projects/adventscala/Day10.scala
package com.nibado.projects.adventscala

class Day10 extends Day {
  override def run(): Unit = {
    val input = getResourceAsString("/day10.txt")
    transform(40, input)
    transform(50, input)
  }

  def transform(times: Int, input: String): Unit = {
    var current = input;
    (1 to times).foreach(n => current = transform(current))
    printAnswer(10, if(times == 40) "One" else "Two", current.length)
  }

  def transform(input: String) : String = {
    val it = input.iterator
    val build = new StringBuilder
    val output = new StringBuilder
    while(it.hasNext) {
      val c = it.next()
      if(build.nonEmpty && build.charAt(0) != c) {
        output.append(build.length.toString).append(build.charAt(0))
        build.setLength(0)
      }
      build += c
    }

    output.append(build.length.toString).append(build.charAt(0)).toString()
  }
}
